rust/worker/src/system/types.rs

```diff
@@ -92,41 +173,55 @@ impl<C: Component> ComponentHandle<C> {
         // and instead use a one shot channel to signal completion
         // TODO: implement this
         join_handle: Option<tokio::task::JoinHandle<()>>,
-        sender: Sender<C>,
+        sender: ComponentSender<C>,
     ) -> Self {
         ComponentHandle {
             cancellation_token: cancellation_token,
-            state: ComponentState::Running,
-            join_handle: join_handle,
+            state: Arc::new(Mutex::new(ComponentState::Running)),
+            join_handle: Arc::new(Mutex::new(
+                join_handle.map(|handle| ConsumableJoinHandle::new(handle)),
+            )),
             sender: sender,
         }
     }
 
-    pub(crate) fn stop(&mut self) {
+    pub(crate) async fn stop(&mut self) {
+        let mut state = self.state.lock().await;
         self.cancellation_token.cancel();
-        self.state = ComponentState::Stopped;
+        *state = ComponentState::Stopped;
     }
 
-    pub(crate) async fn join(&mut self) {
-        match self.join_handle.take() {
-            Some(handle) => {
-                handle.await;
-            }
-            None => return,
-        };
+    /// Consumes the underlying join handle. Panics if it is consumed twice.
+    pub(crate) async fn join(&mut self) -> Result<(), JoinError> {
+        if let Some(join_handle) = self.join_handle.lock().await.as_mut() {
+            join_handle.consume().await
+        } else {
+            Ok(())
+        }
     }
 
-    pub(crate) fn state(&self) -> &ComponentState {
-        return &self.state;
+    pub(crate) async fn get_current_state(&self) -> ComponentState {
+        return self.state.lock().await.clone();
```

Comment:
we can derive(Copy) for state since its an enum and clone is cheap - then avoid the clone

Reply:
what does the syntax look like for this? normally to copy a value I would directly return `self.state.lock()` but that results in an error since `.lock()` returns a mutex guard

Reply:
I think you can deref the guard *? this is nbd if its not straightforward, I'd have to check

Reply:
ah yep, ty